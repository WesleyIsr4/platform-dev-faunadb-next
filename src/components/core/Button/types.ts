import { ReactNode, ButtonHTMLAttributes, ReactElement } from 'react'

import { colors } from 'styles/theme/configs/colors/types'

export type Button = {
  children: ReactNode
  disabled?: boolean
  onClick?: () => void
  small?: boolean
  type?: 'submit' | 'reset' | 'button'
  loading?: boolean
  color?: colors
  fullWidth?: boolean
  variant?: 'primary' | 'secondary' | 'tertiary' | 'text'
  background?: colors
  border?: colors
  leftElement?: ReactElement
  rightElement?: ReactElement
} & ButtonHTMLAttributes<HTMLButtonElement>
