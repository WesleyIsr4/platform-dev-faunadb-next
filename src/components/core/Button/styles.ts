import styled, { css, DefaultTheme } from 'styled-components'
import { shade } from 'polished'

import * as T from './types'

const modifiers: any = {
  disabled: () => css`
    cursor: not-allowed;
  `,
  fullWidth: () => css`
    width: 100%;
  `,
  variant: {
    primary: (theme: DefaultTheme, disabled?: boolean) => css`
      background: ${theme.colors.primaryDarkest};
      color: ${theme.colors.neutralLightest};
      font-weight: ${theme.fonts.weight.semiBold};

      &:hover,
      :focus {
        background: ${!disabled && shade(0.2, theme.colors.primaryDark)};
      }

      &:active {
        background: ${!disabled && theme.colors.secondaryDarkest};
      }
    `,
    secondary: (theme: DefaultTheme, disabled?: boolean) => css`
      border: 1px solid ${theme.colors.primaryDarkest};
      background: none;
      color: ${theme.colors.neutralLightest};

      &:hover,
      :focus {
        border: 1px solid ${!disabled && shade(0.5, theme.colors.primaryDark)};
      }

      &:active {
        background: ${!disabled && theme.colors.neutralDark};
      }
    `,
    tertiary: (theme: DefaultTheme) => css`
      background: transparent;
      color: ${theme.colors.neutralBase};
      border: 1px solid ${theme.colors.secondaryDark};
    `,
    text: (theme: DefaultTheme, disabled?: boolean) => css`
      border: none;
      background: none;
      font-weight: ${theme.fonts.weight.semiBold};
      color: ${theme.colors.neutralLightest};

      &:hover,
      :focus {
        color: 1px solid ${!disabled && shade(0.5, theme.colors.neutralLight)};
      }

      &:active {
        background: ${!disabled && theme.colors.neutralDark};
      }
    `,
  },
}

export const Button = styled.button<T.Button>`
  cursor: pointer;
  font-weight: 400;
  border: none;
  border-radius: 40px;
  display: flex;
  justify-content: center;
  align-items: center;
  gap: 16px;
  white-space: nowrap;
  height: 56px;
  font-size: 18px;
  padding: 13px 24px;
  transition: ${({ theme }) => theme.helpers.transitions(['background', 'opacity'])};

  ${({ theme }) => theme.breakPoints.lessThan.tablet} {
    gap: 8px;
  }

  ${({ fullWidth }) => fullWidth && modifiers.fullWidth()};
  ${({ disabled }) => disabled && modifiers.disabled()};
  ${({ theme, disabled, variant = 'primary' }) => modifiers.variant[variant](theme, disabled)};
`
