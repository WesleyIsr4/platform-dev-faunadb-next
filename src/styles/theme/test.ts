import * as theme from './index'

describe('theme', () => {
  it('Theme should match snapshot', () => {
    expect(theme).toMatchSnapshot()
  })

  it('transition helper should return correct string', () => {
    expect(theme.helpers.transitions('background')).toBe('background 400ms ease')
  })
})
