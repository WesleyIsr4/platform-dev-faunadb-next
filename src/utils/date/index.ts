import { format } from 'date-fns'

import type * as T from './types'

const DEFAULT_FORMAT = 'yyyy/MM/dd'

export const normalizeDate = (value: any, maskFormat: T.Format = DEFAULT_FORMAT) => {
  return format(new Date(value), maskFormat)
}
