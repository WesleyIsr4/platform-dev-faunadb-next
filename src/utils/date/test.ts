import { normalizeDate } from './index'

describe('date utils', () => {
  it('normalizeDate', () => {
    expect(normalizeDate(`10/30/2022`)).toStrictEqual('2022/10/30')
  })
})
