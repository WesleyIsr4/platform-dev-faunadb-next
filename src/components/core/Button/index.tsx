import { Fragment } from 'react'

import TextLoading from 'components/TextLoading'

import * as S from './styles'
import * as T from './types'

function Button({
  children,
  loading,
  disabled,
  type = 'button',
  variant = 'primary',
  color,
  leftElement,
  rightElement,
  background,
  border,
  ...rest
}: T.Button) {
  return (
    <S.Button
      {...rest}
      border={border}
      color={color}
      background={background}
      disabled={disabled || loading}
      type={type}
      variant={variant}
    >
      {loading ? (
        <TextLoading>
          {leftElement}
          {children}
          {rightElement}
        </TextLoading>
      ) : (
        <Fragment>
          {leftElement}
          {children}
          {rightElement}
        </Fragment>
      )}
    </S.Button>
  )
}

export default Button
