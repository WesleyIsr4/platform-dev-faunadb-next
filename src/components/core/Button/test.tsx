import { render, screen } from 'utils/test'

import Button from './index'

describe('<Button />', () => {
  it('should button match snapshot', () => {
    const { container } = render(<Button>I am a Button</Button>)

    expect(container).toMatchSnapshot()
  })

  it('should render the default button condition', () => {
    render(<Button>I am a Button</Button>)

    expect(screen.getByRole('button', { name: /I am a Button/i })).toHaveStyle({
      backgroundColor: '#5F2EEA',
      padding: '13px 24px',
      height: '56px',
      fontSize: '18px',
    })
  })

  it('Should render secondary variant', () => {
    render(<Button variant="secondary">Secondary Button</Button>)

    expect(screen.getByRole('button', { name: /Secondary Button/i })).toHaveStyle({
      background: 'none',
    })
  })

  it('Should render tertiary variant', () => {
    render(<Button variant="tertiary">Tertiary Button</Button>)

    expect(screen.getByRole('button', { name: /Tertiary Button/i })).toHaveStyle({
      backgroundColor: 'transparent',
      border: '1px solid #0096B7',
    })
  })

  it('Should render text variant', () => {
    render(<Button variant="text">Text Button</Button>)

    expect(screen.getByRole('button', { name: /Text Button/i })).toHaveStyle({
      background: 'none',
      border: 'none',
      fontWeight: 500,
      color: '#F1F1F1',
    })
  })
})
