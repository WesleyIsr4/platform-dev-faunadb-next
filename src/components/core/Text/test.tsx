import { render, screen } from 'utils/test'

import Text from './index'

describe('<Text />', () => {
  it('should render correct color', () => {
    render(<Text color="primaryLight">Text with primary light color</Text>)

    expect(screen.queryByText('Text with primary light color')).toHaveStyle({ color: '#E4DAFF' })
  })
})
