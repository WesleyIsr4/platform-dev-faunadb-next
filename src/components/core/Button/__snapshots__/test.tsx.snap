// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should button match snapshot 1`] = `
.c0 {
  cursor: pointer;
  font-weight: 400;
  border: none;
  border-radius: 40px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 16px;
  white-space: nowrap;
  height: 56px;
  font-size: 18px;
  padding: 13px 24px;
  -webkit-transition: background 400ms ease,opacity 400ms ease;
  transition: background 400ms ease,opacity 400ms ease;
  background: #5F2EEA;
  color: #F1F1F1;
  font-weight: 500;
}

.c0:hover,
.c0:focus {
  background: #210081;
}

.c0:active {
  background: #1CC8EE;
}

@media screen and (max-width:768px) {
  .c0 {
    gap: 8px;
  }
}

<div>
  <button
    class="c0"
    type="button"
  >
    I am a Button
  </button>
</div>
`;
